errors.go

```diff
@@ -80,3 +81,125 @@ var (
+	ErrorTypePermission                                  // insufficient permissions
+	ErrorTypeParameterNotResolved                        // parameter types could not be resolved
+	ErrorTypeParameterNotAllowed                         // parameter types not allowed
+	ErrorTypeDependency                                  // dependency
+	ErrorTypeHTTP
+	ErrorTypeMissingExtension // Thrown when an extension is used but not loaded
+	ErrorTypeAutoLoad         // Thrown when an extension is used but not loaded
+	ErrorTypeSequence
+)
+
+var exceptionPrefixMap = map[string]DuckDBErrorType{
+	"Invalid Error":                ErrorTypeInvalid,
+	"Out of Range Error":           ErrorTypeOutOfRange,
+	"Conversion Error":             ErrorTypeConversion,
+	"Error":                        ErrorTypeUnknownType,
+	"Decimal Error":                ErrorTypeDecimal,
+	"Mismatch Type Error":          ErrorTypeMismatchType,
+	"Divide by Zero Error":         ErrorTypeDivideByZero,
+	"Object Size Error":            ErrorTypeObjectSize,
+	"Invalid type Error":           ErrorTypeInvalidType,
+	"Serialization Error":          ErrorTypeSerialization,
+	"TransactionContext Error":     ErrorTypeTransaction,
+	"Not implemented Error":        ErrorTypeNotImplemented,
+	"Expression Error":             ErrorTypeExpression,
+	"Catalog Error":                ErrorTypeCatalog,
+	"Parser Error":                 ErrorTypeParser,
+	"Planner Error":                ErrorTypePlanner,
+	"Scheduler Error":              ErrorTypeScheduler,
+	"Executor Error":               ErrorTypeExecutor,
+	"Constraint Error":             ErrorTypeConstraint,
+	"Index Error":                  ErrorTypeIndex,
+	"Stat Error":                   ErrorTypeStat,
+	"Connection Error":             ErrorTypeConnection,
+	"Syntax Error":                 ErrorTypeSyntax,
+	"Settings Error":               ErrorTypeSettings,
+	"Binder Error":                 ErrorTypeBinder,
+	"Network Error":                ErrorTypeNetwork,
+	"Optimizer Error":              ErrorTypeOptimizer,
+	"NullPointer Error":            ErrorTypeNullPointer,
+	"IO Error":                     ErrorTypeIO,
+	"INTERRUPT Error":              ErrorTypeInterrupt,
+	"FATAL Error":                  ErrorTypeFatal,
+	"INTERNAL Error":               ErrorTypeInternal,
+	"Invalid Input Error":          ErrorTypeInvalidInput,
+	"Out of Memory Error":          ErrorTypeOutOfMemory,
+	"Permission Error":             ErrorTypePermission,
+	"Parameter Not Resolved Error": ErrorTypeParameterNotResolved,
+	"Parameter Not Allowed Error":  ErrorTypeParameterNotAllowed,
+	"Dependency Error":             ErrorTypeDependency,
+	"HTTP Error":                   ErrorTypeHTTP,
+	"Missing Extension Error":      ErrorTypeMissingExtension,
+	"Extension Autoloading Error":  ErrorTypeAutoLoad,
+	"Sequence Error":               ErrorTypeSequence,
+}
+
+type DuckDBError struct {
+	Type DuckDBErrorType
+	Msg  string
+}
```

Comment:
I think dirver errors and errors from db engine is actually different things. If we take a look at the [go-mysql driver](https://github.com/go-sql-driver/mysql/blob/master/packets.go#L584), they only use type `MySQLError` to represent an error comes from the MySQL engine(normaly a SQL syntax error or just some tables/fields does not exist), so we are better to make db errors and driver errors be splited types, like `GoDuckDBError` and `DriverError`(but look below).

> I am asking this because ideally go-duckdb does not panic on driver errors like the one below.
As I know, using a closed connection is a deadly failure, on Unix/Linux it always sends a bad pipe signal and the c/c++ program crashes, so panicking on this error won’t surprise the user. A driver error can come from misusing, incorrect configurations or environment problems(etc. network), even if these errors are returned, users usually cannot take other actions to effectively solve the problem except outputting logs. My personal opinion is let these errors panic is the better way.

Reply:
These are fair points - let's stick to introducing `DuckDBError` in this PR.